wandb/sdk/launch/create_job.py

```diff
@@ -0,0 +1,430 @@
+            runtime=runtime,
+        )
+        if not repo_metadata:
+            tempdir.cleanup()  # otherwise git can pollute
+            return None, "", []
+        metadata.update(repo_metadata)
+    elif job_type == "code":
+        path, entrypoint = _handle_artifact_entrypoint(path, entrypoint)
+        if not entrypoint:
+            wandb.termerror(
+                "Artifact jobs must have an entrypoint, either included in the path or specified with -E"
+            )
+            return None, "", []
+
+        artifact_metadata, requirements = _create_artifact_metadata(
+            path=path, entrypoint=entrypoint, runtime=runtime
+        )
+        if not artifact_metadata:
+            return None, "", []
+        metadata.update(artifact_metadata)
+    elif job_type == "image":
+        if runtime:
+            wandb.termwarn("Runtime is not supported for image jobs, ignoring runtime")
+        if entrypoint:
+            wandb.termwarn(
+                "Entrypoint is not supported for image jobs, ignoring entrypoint"
+            )
+        metadata.update({"python": runtime or "", "docker": path})
+    else:
+        wandb.termerror(f"Invalid job type: {job_type}")
+        return None, "", []
+
+    dump_metadata_and_requirements(
+        metadata=metadata,
+        tmp_path=tempdir.name,
+        requirements=requirements,
+    )
+
+    # init hidden wandb run with job building disabled (handled manually)
+    run = wandb.init(
+        dir=tempdir.name,
+        settings={"silent": True, "disable_job_creation": True},
+        entity=entity,
+        project=project,
+        job_type="cli_create_job",
+    )
+    job_builder = _configure_job_builder_for_partial(tempdir.name, job_source=job_type)
+    if job_type == "code":
+        full_path = os.path.join(path, entrypoint or "")
+        artifact_name = make_code_artifact_name(full_path, name)
+        code_artifact = wandb.Artifact(
+            name=artifact_name,
+            type="code",
+            description="Code artifact for job",
+        )
+        code_artifact.add_dir(path)
+        res, _ = api.create_artifact(
+            artifact_type_name="code",
+            artifact_collection_name=artifact_name,
+            digest=code_artifact.digest,
+            client_id=code_artifact._client_id,
+            sequence_client_id=code_artifact._sequence_client_id,
+            entity_name=entity,
+            project_name=project,
+            run_name=run.id,  # run will be deleted after creation
+            description="Code artifact for job",
+            metadata={"codePath": path, "entrypoint": entrypoint},
+            is_user_created=True,
+            aliases=[
+                {"artifactCollectionName": artifact_name, "alias": a} for a in aliases
+            ],
+        )
+        run.log_artifact(code_artifact)
+        code_artifact.wait()
+        job_builder._set_logged_code_artifact(res, code_artifact)
+        if name:
```

Comment:
Wait, shouldn't we just use their name if they passed a name?

Reply:
Because its the code artifact, we append "code" to whatever the user passed in, then remove it to make it look like exactly what they wanted when constructing the actual job.

Reply:
The name can't be the exact same, as they would collide, so we always do **some** manipulation of the name. It looks gross but works reliably 🤷🏼‍♂️

Reply:
If the name had illegal characters that we formatted with `make_artifact_name_safe` we aren't guaranteed those are the same.

Reply:
Not sure that *actually* matters, but it would break from our current naming scheme where the code artifact job is exactly the job name with `code` instead of `job`...

Reply:
Got it. Okay, so the code artifact name is already safe. Ideally, if the user passes an invalid name, we check it early and throw an error before making any mutations.

Reply:
Yeah ez, adding